net/dns/quad100.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package dns
+
+import (
+	"bytes"
+	"context"
+	"net"
+	"net/netip"
+	"time"
+)
+
+type ManagerConn struct {
+	Ctx        context.Context
+	DnsManager *Manager
+
+	rbuf bytes.Buffer
+}
+
+var (
+	_ net.Conn       = (*ManagerConn)(nil)
+	_ net.PacketConn = (*ManagerConn)(nil) // be a PacketConn to change net.Resolver semantics
+)
+
+func (*ManagerConn) Close() error                       { return nil }
+func (*ManagerConn) LocalAddr() net.Addr                { return todoAddr{} }
+func (*ManagerConn) RemoteAddr() net.Addr               { return todoAddr{} }
+func (*ManagerConn) SetDeadline(t time.Time) error      { return nil }
+func (*ManagerConn) SetReadDeadline(t time.Time) error  { return nil }
+func (*ManagerConn) SetWriteDeadline(t time.Time) error { return nil }
+
+func (c *ManagerConn) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return c.Write(p)
+}
+
+func (c *ManagerConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	n, err = c.Read(p)
+	return n, todoAddr{}, err
+}
+
+func (c *ManagerConn) Read(p []byte) (n int, err error) {
+	return c.rbuf.Read(p)
+}
+
+func (c *ManagerConn) Write(packet []byte) (n int, err error) {
+	pkt, err := c.DnsManager.Query(c.Ctx, packet, "tcp", netip.AddrPort{})
+	if err != nil {
+		return 0, err
+	}
+	c.rbuf.Write(pkt)
+	return len(packet), nil
+}
```

Comment:
sure these Read and Write on rbuf are never called concurrently?

We might want a buffered pipe here instead of bytes.Buffer

Reply:
This is based off the code in `net/tsdial/dohclient.go`, it seems that a buffered pipe isn't used there so this was kept as is :shrug: